horovod/common/basics.py

```diff
@@ -23,22 +23,32 @@ class Comm:
         ...
 
 from horovod.common.process_sets import ProcessSet, global_process_set, _init_process_sets
+from horovod.common.process_sets import _setup as _setup_process_sets
 from horovod.common import util as util
 
 
 class HorovodBasics(object):
     """Wrapper class for the basic Horovod API."""
 
     def __init__(self, pkg_path, *args):
-        full_path = util.get_extension_full_path(pkg_path, *args)
-        self.MPI_LIB_CTYPES = ctypes.CDLL(full_path, mode=ctypes.RTLD_GLOBAL)
-
-        self.Average = self.MPI_LIB_CTYPES.horovod_reduce_op_average()
-        self.Sum = self.MPI_LIB_CTYPES.horovod_reduce_op_sum()
-        self.Adasum = self.MPI_LIB_CTYPES.horovod_reduce_op_adasum()
-        self.Min = self.MPI_LIB_CTYPES.horovod_reduce_op_min()
-        self.Max = self.MPI_LIB_CTYPES.horovod_reduce_op_max()
-        self.Product = self.MPI_LIB_CTYPES.horovod_reduce_op_product()
+        self.full_path = util.get_extension_full_path(pkg_path, *args)
+
+        # These ReduceOp must be kept in sync with horovod/common/message.h
+        # We can not get these values from ctypes (self.MPI_LIB_CTYPES.horovod_reduce_op_average() etc.) dynamically
+        # because ctypes hasn't load shared lib yet, but other components need to import these ReduceOp now.
+        AVERAGE = 0
+        SUM = 1
+        ADASUM = 2
+        MIN = 3
+        MAX = 4
+        PRODUCT = 5
```

Comment:
Please put these definition outside of this class's init function. Can also be out of the class.